File: Niveau 4/2 – Arbres/5 - Anti-virus.py
def checkMask(element, mask):
  element = str(element)
  if(len(element) == len(mask)):
    for i in range(len(element)):
      if(mask[i] != '?' and mask[i] != element[i]):
        return False
    return True
  return False
def main():
  nbProduits = int(input())
  produits = list(map(int, input().split()))
  tree = [[] for i in range(nbProduits + 1)]
  for i in range(nbProduits):
    tree[produits[i]].append(i + 1)
  mask = input()
  nodes = [0]
  while(len(nodes) != 0):
    currentNode = nodes.pop(0)
    if(currentNode != 0 and checkMask(currentNode, mask)):
      print(currentNode, end=" ")
    for i in range(len(tree[currentNode])):
      child = tree[currentNode][i]
      nodes.append(child)

main()

File: Niveau 4/2 – Arbres/5 - Anti-virus.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

bool checkMask(int element, string mask)
{
    string strElement = to_string(element);
    if (strElement.size() == mask.size())
    {
        for (size_t i = 0; i < strElement.size(); i++)
        {
            if (mask[i] != '?' && mask[i] != strElement[i])
            {
                return false;
            }
        }
        return true;
    }
    return false;
}

int main()
{
    ios::sync_with_stdio(false);
    int n;
    vector<int> produits;
    cin >> n;
    vector<vector<int>> tree(n + 1);
    for (int i = 0; i < n; i++)
    {
        int element;
        cin >> element;
        produits.push_back(element);
    }
    for (int i = 0; i < n; i++)
    {
        tree[produits[i]].push_back(i + 1);
    }
    string mask;
    cin >> mask;
    queue<int> q;
    q.push(0);
    while (!q.empty())
    {
        int currentNode = q.front();
        q.pop();
        if (currentNode != 0 and checkMask(currentNode, mask))
        {
            cout << currentNode << ' ';
        }
        for (int child : tree[currentNode])
        {
            q.push(child);
        }
    }
}